### ISSUE TITLE
MySQL: sqlfluff stuck in parse stage when the file contains long list of values

### ISSUE THREAD

**Issue by ngdah** [sqlfluff/sqlfluff#4385]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

we have a file with very long value list, sqlfluff took super long (tens of minutes) to lint.

### Expected Behaviour

it should finish in seconds

### Observed Behaviour

the command took very long in parser and lint rules

### How to reproduce

command to run:
```
sqlfluff lint --dialect mysql file.sql
```
use the file below:
```
DELETE FROM
    db_name.table_name
WHERE id IN (
    2129, 1933, 2300, 1810, 1317, 2564, 2757, 1489, 2194, 1477, 2186, 2156, 2202, 2344, 1788, 1842, 2043, 2534, 1798, 2036,
    1736, 2079, 1993, 1727, 1926, 1463, 1692, 2628, 2820, 1915, 2448, 1525, 2578, 2770, 2277, 1648, 1860, 1861, 2659, 2851,
    2536, 2108, 2319, 2239, 1914, 1743, 2353, 1321, 2565, 2758, 1899, 2198, 2148, 2691, 2416, 1296, 1407, 1531, 2307, 2514,
    2744, 1994, 1677, 1739, 2093, 1353, 2568, 2761, 1905, 2666, 2858, 1645, 2618, 2810, 2291, 2188, 2699, 1895, 1456, 1951,
    2246, 1997, 1306, 2562, 2755, 1533, 2005, 1403, 1970, 2478, 1330, 1825, 1941, 2465, 2741, 1707, 2399, 1617, 1435, 1335,
    1653, 2621, 2813, 2380, 1480, 1634, 2611, 2803, 1896, 2189, 1831, 2173, 1624, 2223, 2214, 1439, 2032, 2144, 2265, 2711,
    2272, 2713, 1469, 1464, 2164, 1421, 1764, 2646, 2838, 2090, 1950, 2401, 2234, 2474, 1336, 1957, 1793, 2649, 2841, 2487,
    1678, 2625, 2817, 2372, 1393, 2331, 1311, 1520, 1790, 2503, 1694, 2629, 2821, 1829, 1932, 1462, 1526, 1511, 1850, 1839,
    2386, 1404, 2305, 2103, 2349, 2722, 1443, 1389, 2476, 1982, 1631, 2140, 1721, 2507, 1626, 2175, 1639, 2615, 2807, 1777,
    2072, 1510, 2253, 1551, 2010, 2674, 2163, 1995, 1706, 2633, 2825, 1628, 2608, 2800, 1674, 1821, 2656, 2848, 2000, 2512,
    1733, 2641, 2833, 1297, 1728, 2640, 2832, 1377, 2258, 1378, 1757, 2643, 2835, 2160, 2694, 1769, 2011, 1918, 2472, 2034,
    2678, 2523, 1535, 1332, 2409, 1397, 2132, 2178, 1863, 2660, 2852, 1767, 1906, 1670, 1578, 1724, 2210, 2703, 1676, 2624,
    2816, 2280, 2444, 1732, 1562, 1308, 1959, 2321, 2350, 2471, 2052, 1299, 2151, 1833, 2412, 2101, 2373, 1760, 2106, 2123,
    1804, 1379, 2022, 1457, 1319, 2393, 2730, 2166, 1746, 1802, 2651, 2843, 1908, 1471, 2367, 2450, 2228, 2023, 2485, 2499,
    2351, 2723, 1787, 1454, 1619, 2602, 2794, 1827, 2658, 2850, 2429, 1475, 2451, 2038, 2048, 2181, 1566, 2356, 2724, 2097,
    2301, 2328, 2415, 2247, 1342, 2159, 1425, 1956, 1423, 1438, 2571, 2764, 1695, 2630, 2822, 2063, 1583, 2366, 2113, 1557,
    2303, 2320, 1334, 2218, 2318, 2337, 2439, 2734, 1494, 2543, 1506, 1973, 1622, 2605, 2797, 2548, 1558, 2250, 1864, 2126,
    1427, 1776, 2074, 1719, 2021, 1847, 1819, 1371, 2546, 1320, 1939, 1688, 1837, 1735, 2385, 1998, 2243, 1552, 2383, 2530,
    1666, 1595, 2376, 2347, 1573, 2003, 2068, 2683, 1942, 2069, 1380, 2335, 1440, 1680, 2098, 2545, 2185, 2398, 1387, 1636,
    2613, 2805, 2075, 2422, 2092, 1581, 1931, 2668, 2860, 1758, 1752, 1430, 1922, 1917, 1478, 2572, 2765, 2076, 2323, 1884,
    2229, 2705, 1363, 1682, 1561, 2137, 2362, 1384, 2520, 2264, 2025, 1591, 2446, 2058, 1775, 1614, 1741, 1893, 2139, 1633,
    1990, 2670, 2862, 2486, 2124, 1556, 1949, 2041, 1999, 2039, 2679, 2172, 1923, 1929, 1865, 2297, 2343, 1345, 1375, 1406,
    1338, 1659, 2316, 2718, 1961, 1979, 2191, 1448, 1607, 2595, 2787, 1650, 2620, 2812, 2555, 1781, 1713, 2635, 2827, 2493,
    1960, 1627, 2294, 1657, 2206, 2267, 2336, 1877, 2662, 2854, 2315, 1902, 1938, 2431, 2127, 2689, 2325, 2286, 2287, 2714,
    2715, 2256, 1768, 1479, 2027, 2274, 2500, 2271, 1391, 1548, 2091, 1615, 2601, 2793, 2053, 2120, 2384, 2378, 1568, 2237,
    2295, 2405, 2731, 1484, 2088, 1714, 2107, 2687, 1681, 1797, 1541, 2284, 1763, 1980, 2306, 1771, 2014, 1577, 2584, 2776,
    1664, 1452, 1808, 1989, 2238, 2533, 1487, 1474, 1482, 1383, 1350, 1963, 1340, 1691, 1838, 2538, 2544, 2749, 1885, 2664,
    2856, 2016, 1328, 2455, 1873, 1946, 2524, 2227, 2488, 2207, 2702, 1547, 2580, 2772, 2312, 1412, 1314, 1805, 1754, 2054,
    1401, 2033, 1859, 1755, 2642, 2834, 1524, 1592, 1697, 2553, 1794, 1900, 1953, 2042, 2516, 2060, 2293, 2389, 1392, 1609,
    1492, 2109, 1599, 2588, 2780, 1356, 2552, 1343, 1372, 2463, 2391, 2049, 2680, 1856, 1587, 1872, 2661, 2853, 2308, 2355,
    2359, 2390, 2729, 2402, 1928, 1597, 2586, 2778, 2363, 1349, 1846, 2438, 1574, 2125, 1828, 1806, 2653, 2845, 2008, 1962,
    2669, 2861, 2184, 2015, 2454, 2738, 1888, 2013, 1413, 1920, 1730, 1801, 2375, 2045, 2279, 1910, 1921, 2667, 2859, 1536,
    2379, 1916, 2285, 1641, 1346, 1542, 1331, 1876, 1417, 1359, 1466, 2494, 1673, 1765, 2200, 1527, 2169, 1835, 1742, 2217,
    1589, 2365, 1925, 1352, 1807, 1625, 2607, 2799, 2086, 2381, 1313, 1987, 2017, 2195, 1348, 2522, 1968, 1623, 2606, 2798,
    1483, 1832, 1803, 2652, 2844, 2233, 2707, 1537, 1422, 1965, 1588, 1661, 2622, 2814, 1992, 1414, 1642, 1575, 2583, 2775,
    1394, 2310, 1339, 2566, 2759, 2509, 1455, 2466, 2508, 1470, 2066, 2358, 1845, 2146, 1386, 2426, 1415, 1879, 2663, 2855,
    2121, 2510, 1761, 2645, 2837, 2215, 1364, 1685, 1434, 2281, 2212, 2141, 1337, 1420, 1385, 2569, 2762, 1651, 2434, 2535,
    1930, 1785, 2648, 2840, 1501, 2576, 2768, 2204, 2701, 1309, 1461, 1947, 1710, 2634, 2826, 2556, 2528, 1621, 2604, 2796,
    2436, 2492, 1523, 2577, 2769, 2130, 1481, 1762, 2216, 2704, 2490, 2176, 2219, 2057, 1613, 2600, 2792, 1853, 1662, 2224,
    1671, 2623, 2815, 1909, 2406, 2527, 2423, 1485, 2473, 1978, 1686, 2004, 2672, 2346, 1593, 2112, 2158, 2245, 2174, 2697,
    2055, 1871, 2192, 1539, 1655, 1502, 2089, 2470, 1540, 2360, 2725, 1894, 1855, 2119, 1668, 2018, 2073, 2684, 1799, 1649,
    2619, 2811, 2526, 1298, 2388, 1512, 1988, 1373, 1601, 2589, 2781, 1500, 2542, 2748, 1465, 1704, 2222, 1532, 1744, 2031,
    1519, 2425, 2254, 1996, 2671, 2071, 2213, 2460, 2467, 1398, 2400, 1840, 2475, 1774, 2377, 1428, 1722, 2638, 2830, 1432,
    2529, 2746, 1972, 2128, 1843, 1907, 2443, 2432, 1717, 2636, 2828, 1307, 1402, 1841, 2050, 2251, 2709, 2498, 2299, 2717,
    2410, 1498, 2575, 2767, 1783, 2067, 2232, 1751, 2292, 1854, 1376, 2452, 2506, 2456, 1606, 2594, 2786, 2110, 2403, 1984,
    2059, 1458, 1886, 2665, 2857, 1849, 1576, 1509, 2424, 1815, 1545, 1690, 2627, 2819, 1504, 1362, 1514, 2081, 1824, 2111,
    2688, 1816, 1495, 2574, 2766, 2352, 2268, 2428, 1447, 2304, 1312, 2077, 2138, 2483, 2502, 1976, 2329, 2193, 1361, 2269,
    2712, 1486, 2339, 2322, 1491, 2196, 2700, 2134, 2690, 1822, 1658, 2157, 1887, 1418, 1977, 1779, 2205, 2203, 1780, 2012,
    2168, 2696, 2078, 2221, 1351, 1745, 2539, 2040, 2201, 1981, 2149, 1400, 1952, 1934, 1300, 1789, 1737, 1305, 2220, 2282,
    2161, 2342, 1940, 2155, 2693, 2046, 2030, 2677, 1766, 2208, 1582, 2479, 2743, 2087, 2340, 1429, 1958, 2179, 1883, 1820,
    2094, 2686, 1867, 1424, 2570, 2763, 2133, 1705, 2632, 2824, 2260, 2298, 2211, 2122, 1590, 2518, 2745, 1366, 1919, 2240,
    2427, 2117, 2440, 2505, 1702, 1629, 2609, 2801, 1675, 1543, 2579, 2771, 2330, 1565, 1517, 1795, 1836, 1518, 1954, 1715,
    1503, 2037, 1468, 2557, 2750, 2480, 1460, 2235, 1725, 2374, 1753, 1529, 2236, 2708, 2290, 1898, 2131, 2102, 2020, 2521,
    2064, 1862, 1759, 2644, 2836, 1778, 1496, 1304, 1667, 2332, 1515, 1488, 2573, 1499, 2357, 1818, 1687, 1709, 2407, 1333,
    2145, 1549, 1381, 1490, 1436, 1927, 2309, 1712, 2080, 1610, 2597, 2789, 2180, 1390, 1553, 1974, 1748, 2519, 2496, 2199,
    1374, 2541, 2165, 1608, 2596, 2788, 1431, 1646, 1360, 1943, 1696, 1866, 1786, 1878, 2115, 1405, 1618, 2255, 1445, 2408,
    2056, 2170, 2070, 1357, 1811, 1476, 2225, 1985, 1546, 2419, 2732, 1555, 1948, 2007, 2445, 1912, 1935, 1395, 2257, 2551,
    1901, 2241, 2491, 1453, 1315, 2563, 2756, 1530, 2404, 2532, 2114, 1522, 1516, 1800, 1924, 1851, 1937, 1897, 2481, 1370,
    1327, 2513, 2345, 1656, 1880, 1612, 2599, 2791, 1852, 1368, 2152, 2361, 1441, 1665, 1419, 2449, 2736, 2468, 1823, 1647,
    2459, 1564, 1913, 1891, 1936, 2162, 2695, 2369, 2727, 1605, 2593, 2785, 1882, 1369, 2392, 2537, 2747, 1792, 2183, 2698,
    2275, 1329, 1868, 1399, 1408, 1663, 2311, 2019, 2062, 1782, 2504, 2317, 2719, 2001, 1969, 1572, 2368, 1857, 2266, 2051,
    2167, 2417, 1826, 2657, 2849, 2147, 2276, 1975, 1966, 2489, 2394, 2442, 1784, 1693, 2244, 1301, 2559, 2752, 1611, 2598,
    2790, 1889, 2314, 2296, 2716, 1513, 1567, 2458, 1718, 2395, 1310, 1493, 1632, 2610, 2802, 2154, 2692, 2382, 2728, 2061,
    2681, 1791, 1497, 2044, 1858, 2150, 1991, 1983, 1701, 1505, 2525, 2469, 2742, 1459, 2554, 1955, 1848, 1382, 1358, 1604,
    2592, 2784, 1749, 1875, 1528, 2547, 1698, 2326, 1635, 2612, 2804, 1813, 2654, 2846, 1409, 2283, 1869, 1600, 1971, 2029,
    2676, 2153, 1585, 2197, 2190, 1892, 1538, 2370, 2082, 2685, 1586, 1652, 1554, 2581, 2773, 1302, 2560, 2753, 2096, 2116,
    2397, 2414, 1699, 1411, 2099, 1580, 2585, 2777, 1964, 2135, 2035, 2371, 1550, 1594, 2354, 2143, 1347, 1679, 1570, 2582,
    2774, 1630, 1444, 2095, 2433, 1723, 1890, 1426, 1507, 2549, 2457, 2739, 2411, 1316, 2441, 1660, 2313, 1637, 1508, 1944,
    1720, 2637, 2829, 2302, 2259, 1473, 2462, 2740, 2248, 1734, 2333, 1911, 2396, 2100, 2421, 2105, 2413, 1729, 2482, 1598,
    2587, 2779, 1437, 2338, 1986, 1654, 2477, 1602, 2590, 2782, 1738, 2348, 2540, 1747, 1544, 2270, 1773, 1669, 2497, 1367,
    1731, 1904, 1703, 2631, 2823, 1684, 2430, 2142, 1521, 1323, 1689, 1450, 1711, 1672, 1596, 2065, 2682, 1817, 1945, 1534,
    2550, 2028, 2675, 2177, 2341, 1640, 1967, 2278, 1750, 2273, 1472, 1603, 2591, 2783, 1451, 2009, 2104, 2495, 2006, 2673,
    1844, 1571, 1620, 2603, 2795, 2182, 1433, 2084, 1559, 2531, 2511, 1467, 1365, 2288, 1344, 2567, 2760, 1410, 1726, 2639,
    2831, 1416, 2461, 2252, 1756, 1809, 1903, 1355, 1770, 2230, 2706, 2464, 1812, 2501, 1303, 2561, 2754, 1318, 2418, 2261,
    2420, 2733, 1716, 2262, 2047, 2085, 1326, 1638, 2614, 2806, 1396, 2334, 1708, 2171, 1796, 2650, 2842, 1700, 1772, 2647,
    2839, 2118, 1616, 1881, 2324, 2720, 1324, 1354, 2083, 1563, 2558, 2751, 2187, 1579, 2437, 1341, 1814, 2655, 2847, 2231,
    2209, 1830, 2484, 2136, 2435, 1325, 2026, 2226, 1834, 1683, 2626, 2818, 2327, 2721, 1322, 1560, 1874, 2515, 1569, 2002,
    2289, 2364, 2726, 1643, 2616, 2808, 2242, 2387, 1740, 1584, 1449, 2263, 2710, 2447, 2735, 2249, 2453, 2737, 1644, 2617,
    2809, 1446, 2517, 1870, 1388, 2024, 1442
);
```

### Dialect

mysql

### Version

1.4.5 and 2.0.0a4

### Configuration

nil

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by WittierDinosaur** [comment]
Hey @ngdah - we've made some performance improvements to the parser and linter in 2.0.0a5 - can you check what the performance is now?

**Comment by nahuelverdugo** [comment]
Hi @WittierDinosaur, our team has a similar issue with a long list of values for a snowflake query. For example, we have run it with different numbers of values, and the increase seems linear. It is around 1.5 seconds of growth for every 100 values added to the query. 

<img width="590" alt="image" src="https://user-images.githubusercontent.com/11074908/223540110-b3154f72-39e6-4026-8982-42c47517ab3c.png">

We run this piece of code with version `2.0.0a5`:

```python 
import time

import sqlfluff

query = """
create table my_example_table as
select *
from (
    values
    {values}
) as my_data_example (value1, value2, value3, value4, value5)
qualify row_number() over (partition by value1, value2 order by my_data_example is not null desc) = 1
"""

values_format = "\t('value1{a}','value2{b}','value{c}','value{d}','value{e}')"

values = [values_format.format(a=0, b=0, c=0, d=0, e=0)]

f = open("values.log", "w")

for n in range(1, 1000):
    values.insert(0, values_format.format(a=n, b=n, c=n, d=n, e=n) + ",")
    test_query = query.format(values="\n".join(values))
    st = time.time()
    tree = sqlfluff.parse(test_query, dialect="snowflake")
    et = time.time()
    print(f"{n} - {et - st}")
    f.write(f"{et - st}\n")
    
f.close()
```

In our case, we are only interested in the parser, but we are still determining if there is something in the configuration we can tweak to speed up the process.

**Comment by harshach** [comment]
@WittierDinosaur can you please look at the above comment. we are seeing the issue with 2.0.0a5. Anything we can do to help here.

**Comment by peterbud** [comment]
With the latest `sqlfluff` the above example code finishes linting in 5 sec:
```bash
sqlfluff lint test.sql -d mysql  5.21s user 0.35s system 99% cpu 5.581 total
```

Is it still an issue?

**Comment by peterbud** [comment]
Closing due to lack of response
